src/lib/crypto/kimchi_backend/gadgets/ec_group.ml

```diff
@@ -0,0 +1,3965 @@
+   *       2) I' = -I
+   *
+   *   Simplification:
+   *
+   *     Uniformly and randomly select initial accumulator point I, instead of using
+   *     the complicated deterministic process.
+   *
+   *     For a z-bit scalar, there are z unique B points.  Each point also has its
+   *     negative, which we cannot add to itself.  Therefore, in total there are
+   *     2z points that we do not want to select as our initial point nor compute
+   *     as an intermediate A point during scaling.  The probability we select or
+   *     compute one of these points is approx 2z^2/n, where n is the order of the
+   *     elliptic curve group.
+   *
+   *     The probability of selecting a bad point is negligible for our applications
+   *     where z is very small (e.g. 256) and n is very large (e.g. 2^256).  Thus,
+   *     we can simply randomly select the initial accumulator I and the
+   *     double-and-add algorithm will succeed with overwhelming probability.
+   *)
+  let acc, _base =
+    List.foldi scalar ~init:(curve.ia.acc, point) (* (acc, base) *)
+      ~f:(fun i (acc, base) bit ->
+        (* Add: sum = acc + base *)
+        let sum = add (module Circuit) external_checks curve acc base in
+        (* Bounds 1:
+         *   Left input is previous result, so already checked.
+         *   Right input is checked by previous doubling check.
+         *   Initial acc and base are gadget inputs and checked externally.
+         *   Result bounds check below.
+         *)
+        Foreign_field.External_checks.append_bound_check external_checks
+        @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x sum ;
+        Foreign_field.External_checks.append_bound_check external_checks
+        @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y sum ;
+
+        (* Group double: double_base = base + base *)
+        let double_base =
+          match doubles with
+          | None ->
+              let double_base =
+                double (module Circuit) external_checks curve base
+              in
+              (* Bounds 2:
+               *   Input is previous result, so already checked.
+               *   Initial base is gadget inputs and checked externally.
+               *   Result bounds check below.
+               *)
+              Foreign_field.External_checks.append_bound_check external_checks
+              @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x double_base ;
+              Foreign_field.External_checks.append_bound_check external_checks
+              @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y double_base ;
+              double_base
+          | Some doubles ->
+              (* When the base point is public (e.g. the secp256k1 generator) we can
+               * improve performance by having them as precomputed public parameters *)
+              doubles.(i)
+        in
+
+        (* Group add conditionally *)
+        let acc = Affine.if_ (module Circuit) bit ~then_:sum ~else_:acc in
```

Comment:
super cool

Reply:
So, just to make sure. This executes both branches (executes = creates circuit for them), regardless of whether it is 1 or 0, no? Otherwise you're filtering the decomposition which could be a problem.

Reply:
Correct.   Both branches are computed and then the if computes and constrains the output based on the test bit, which is witness data.